homeassistant/core.py

```diff
@@ -1662,7 +1664,12 @@ def async_set(
         domain_index[entity_id] = state
         self._bus.async_fire(
             EVENT_STATE_CHANGED,
-            {"entity_id": entity_id, "old_state": old_state, "new_state": state},
+            {
+                "entity_id": entity_id,
+                "old_state": old_state,
+                "new_state": state,
+                "entity_info": entity_info,
```

Comment:
This is going to get serialized downline for everything looking at state change events which is a lot of things (mobile_app, node red, app daemon, various other add-ons)

Reply:
I think mobile_app is only downloading the states so should be ok there

node red looks like they switched to subscribeEntites so should be ok there now https://github.com/zachowj/node-red-contrib-home-assistant-websocket/blob/24327986dc9701096712b693bc8aa925f0871152/src/homeAssistant/Websocket.ts#L275

app daemon looks like its still listening for all state change events

Reply:
Going to put this on my production and test it out

Reply:
Maybe instead of teach state changes events about entities we can have some type of run time only entity registry that isn't saved to disk that the recorder or other integrations can access.  Something simply that is a wrapper around a `dict[str, RunTimeEntityInfo]`

Reply:
We already have such a global registry, the `entity_sources`. The reason I added a reference of the entity info it to the state changed events was to not introduce a new race where recorder is behind with recording and the information in `entity_sources` is no longer valid when the event is processed.

Realistic cases would be when config entry options are changes (causing all its entities to be removed and then recreated), and if the user renames an entity.

However, adding entity info to the `data` dictionary is clearly far from ideal, I have no idea what I was thinking 🤦

Reply:
Modified such that `State` objects now have a reference to the `entity_info`